Create the OmpSs version from the cpp code

LonelyCat124 opened the pull request:
This should be ready for review now, the code is faster than the CPP OpenMP version as is on the BWD nodes on scafellpike, and match the checksums exactly, so reasonably happy. 

The makefile stuff is a bit weird. Essentially the OmpSs fortran compiler will not compile the various dependencies, so they are compiled with mpifort as the OmpSs version I couldn't build with the OpenMP versions of dl_esm_inf or dl_timer, and then there are specific rules for building the OmpSs code.

arporter commented on benchmarks/nemo/nemolite2d/manual_versions/psykal_cpp/Makefile:
```diff
@@ -15,6 +15,8 @@
 
 include ../common.mk
 
+MCXXFLAGS?="--Wn,-O3,-mtune=native,-finline-limit=50000,-fopt-info-all=gnu_opt_report.txt"
```
Please could you add a comment to say what these flags are for (i.e. presumably they're for the OMPSs compiler?).

LonelyCat124 replied:
They probably aren't needed - I put them there in case someone doesn't load the correct `compiler_setup` (or wants to use one where `MCXXFLAGS` isn't defined), but since they only work in the case that someone is using the gnu compiler anyway thats probably moot - I'll probably remove them.

LonelyCat124 replied:
Removed in latest commits.

LonelyCat124 commented:
Ok, I moved it to a new folder and I believe have set it up appropriately, and tested the output. I also reverted changes to the psykal_cpp makefile so there should be no changes in that folder now I believe.

LonelyCat124 commented:
I've not updated the main README.md as the cpp files don't seem to be mentioned in that at current (@sergisiso is this intended or WIP?)

arporter commented:
> 
> 
> I've not updated the main README.md as the cpp files don't seem to be mentioned in that at current (@sergisiso is this intended or WIP?)

I think we'll be charitable to ourselves and call that WIP (at least it is now that you've pointed out the omission :-) )

LonelyCat124 commented:
Aha I just know that Sergi had said he was working on rewriting the "single file" section of the README and wasn't sure if it was part of that or not.

sergisiso commented:
@arporter @AidanChalk Yes, I updated the readme in the C OpenCL PR which should be ready very soon. The implementation and the performance results it gets look great :)

arporter commented on benchmarks/nemo/nemolite2d/manual_versions/psykal_cpp_ompss/Makefile:
```diff
@@ -0,0 +1,77 @@
+time_step_mod.o: ${COMMON_LIB} inf_lib momentum_mod.o continuity_mod.o \
+                 time_update_mod.o boundary_conditions_mod.o
+
+# Linking the executable needs the -lstdc++ to link also the std lib functions
```
Presumbaly we can remove this commented-out section?

LonelyCat124 replied:
Done

arporter commented on benchmarks/nemo/nemolite2d/manual_versions/psykal_cpp_ompss/Makefile:
```diff
@@ -0,0 +1,77 @@
+# Makefile for manual C++ version of PSyKAl'd NEMOLite2D.
```
"...manual C++ OMPSS version of..." (or however OMPSS is capitalised).

LonelyCat124 replied:
Done

arporter commented on benchmarks/nemo/nemolite2d/manual_versions/psykal_cpp_ompss/nemolite2d.f90:
```diff
@@ -0,0 +1,156 @@
+  use gocean2d_io_mod, only: model_write
+  use gocean_mod,      only: model_write_log, gocean_initialise, &
+                             gocean_finalise
+  !use likwid
```
I think we can get rid of the likwid calls/usage from this file.

LonelyCat124 replied:
Done - The main C++ code may also have these calls as I copied from there @sergisiso

sergisiso replied:
tbh the algorithm layer file should be an exact copy in all psykal_* implementations, we can consider (in another PR) that all Makefiles use the same source file or that each folder have a symbolic link to a copy on the parent folder.

arporter replied:
Yes, I was thinking the same @sergisiso but, as you say, that's not something for this PR. (We could also add a profiling wrapper for the likwid API.)

arporter commented on benchmarks/nemo/nemolite2d/manual_versions/psykal_cpp_ompss/Makefile:
```diff
@@ -0,0 +1,77 @@
+	${MAKE} -C ${INF_DIR} distclean
+	${MAKE} -C ${TIMER_DIR} allclean
+
+summary:
```
Just to check: does this target still work OK with this version of the code?

LonelyCat124 replied:
The output I got was:
```
bash-4.2$ make summary
mpicc\t\tcpu\t
0.27785892E+05  0.57543179E+05  0.50304E-01
```
So I think so? The first line seems to not be correct though.

arporter replied:
Weird. It should all be on one line. I get (Gnu Make 4.1):

    $ make summary 
    gcc		cpu	/bin/sh: 1: ./nemolite2d_ompss.exe: not found

Anyway, it prints the results so it's not a big deal unless @sergisiso has any ideas?

LonelyCat124 replied:
It may be a shell thing, if I do:
```
bash-4.2$ echo -n "${CC}\t\tcpu\t"
mpicc\t\tcpu\tbash-4.2$
```

arporter commented on benchmarks/nemo/nemolite2d/manual_versions/psykal_cpp_ompss/time_step_ompss.cpp:
```diff
@@ -0,0 +1,274 @@
+#include <iostream>
+//#include <chrono>
```
Please rm this line.

LonelyCat124 replied:
Done

arporter commented on benchmarks/nemo/nemolite2d/manual_versions/psykal_cpp_ompss/time_step_ompss.cpp:
```diff
@@ -0,0 +1,274 @@
+
+#define TASK_SIZE 32
+//Each double loop over jj then ji is divided into a new structure.
+//The outer loop, now loops over incrememnts of 32 (initially chosen arbitrarliy), and creates tasks of this size
```
s/incrememnts/increments/ and s/arbitrarliy/arbitrarily/ and please break this line and the long ones below.

Although this comment and the ones below talk about increments of 32, the code seems to be parametrised to use increments of TASK_SIZE (which is nicer). If that's correct then please could you update the comments. Maybe also say where TASK_SIZE is set.

arporter replied:
Duh! I see that TASK_SIZE is right there but thanks for adding the comment anyway :-)

arporter commented on benchmarks/nemo/nemolite2d/manual_versions/psykal_cpp_ompss/time_step_ompss.cpp:
```diff
@@ -0,0 +1,274 @@
+//The outer loop, now loops over incrememnts of 32 (initially chosen arbitrarliy), and creates tasks of this size
+//For the dependencies supplied to OmpSs, as opposed to putting the full memory for now, we compute the "task row"
+//For the nth task for each jj+=32 loop, this is n.
+//Tasks read or write to the task_rows appropriately (always either task_row, task_row-1, or task_row+1)
```
according to their stencil accesses? Presumably they always *write* to `task_row`?

arporter commented on benchmarks/nemo/nemolite2d/manual_versions/psykal_cpp_ompss/time_step_ompss.cpp:
```diff
@@ -0,0 +1,274 @@
+      int stop = std::min(jj+TASK_SIZE-1, internal_ystop);
+      int task_row = jj/TASK_SIZE;
+      #pragma omp task firstprivate(jj, task_row, stop) \
+                output(ssha_t[task_row]) input(sshn_u[task_row], \
```
I confess I know very little about tasks and even less about OmpSs but this notation confuses me - how come I only see dependencies in terms of `task_row [+-1]` and not also `task_row [+-TASK_SIZE]` (i.e. for when a kernel reads from locations in other rows)?

LonelyCat124 replied:
The dependencies don't need to actually cover the memory addresses used, so I use task_row as a proxy for groups of TASK_SIZE number of rows. This keeps the soundness the same, but I'd expect is easier for the dependence analysis than groups of multiple rows with sizes attached, and makes the code significantly easier to read.

I could test with the full memory addresses but I changed from doing that to this pretty early in the implementation.

arporter replied:
Aha! Please could you add a comment describing this?

LonelyCat124 replied:
Yeah, I'll extend the comment above to include it.

arporter commented on benchmarks/nemo/nemolite2d/manual_versions/psykal_cpp_ompss/time_step_ompss.cpp:
```diff
@@ -0,0 +1,274 @@
+      }
+    }
+
+#pragma omp taskwait
```
Please add a comment explaining why this wait is required.

LonelyCat124 replied:
Done

arporter commented on compiler_setup/ompss.sh:
```diff
@@ -0,0 +1,36 @@
+#F90FLAGS += -fcheck=all -fbacktrace -ffpe-trap=invalid -g
+F90FLAGS+=" -faggressive-function-elimination"
+F90FLAGS+=$OPTFLAGS
+# f2py does not break long lines so tell gfortran not to
```
You can delete these 3 lines as we do now break lines where appropriate.

LonelyCat124 replied:
Done

arporter replied:
They're still there...?

LonelyCat124 replied:
Yup I removed the wrong line, will be fixed in next commit.

LonelyCat124 commented:
I think these are all done so set it back to ready to review :)

arporter commented on benchmarks/nemo/nemolite2d/manual_versions/psykal_cpp_ompss/Makefile:
```diff
@@ -1,14 +1,8 @@
-# Makefile for manual C++ version of PSyKAl'd NEMOLite2D.
+# Makefile for manual C++ OmpSs version of PSyKAl'd NEMOLite2D.
 #
 # This Makefile picks up the compiler to use any flags from
```
Should this now say use `../../../../../compiler_setup/ompss.sh`?

LonelyCat124 replied:
Done

arporter commented on benchmarks/nemo/nemolite2d/manual_versions/psykal_cpp_ompss/time_step_ompss.cpp:
```diff
@@ -55,10 +54,19 @@ extern "C" void c_invoke_time_step(
 //Each double loop over jj then ji is divided into a new structure.
-//The outer loop, now loops over incrememnts of 32 (initially chosen arbitrarliy), and creates tasks of this size
-//For the dependencies supplied to OmpSs, as opposed to putting the full memory for now, we compute the "task row"
+//The outer loop, now loops over increments of 32 (initially chosen arbitrarily),
```
s/32/TASK_SIZE/ and on line 61?

LonelyCat124 replied:
Done

LonelyCat124 commented:
I think I got them all this time!

arporter commented:
All looks good now thanks.
